File: DigitalImageProcessing/python/convert_color.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 色彩空间的转换; 调整彩色图像的饱和度和亮度
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-21
"""


import sys
import numpy as np
import cv2 as cv


# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1])
        if image is None:
            print(f"Error: no such file or directory.")
            sys.exit()
        cv.imshow("OriginImage", image)

        img_float = image.astype(np.float32)
        image = img_float / 255.0
        HLS_Img = cv.cvtColor(image, cv.COLOR_BGR2HLS)

        l_portion = 0
        s_portion = 0
        MAX_VALUE = 100
        cv.namedWindow("l and s", cv.WINDOW_AUTOSIZE)
        def nothing(*arg):
            pass
        cv.createTrackbar("l", "l and s", l_portion, MAX_VALUE, nothing)
        cv.createTrackbar("s", "l and s", s_portion, MAX_VALUE, nothing)

        lsImg = np.zeros(image.shape, np.float32)
        while True:
            hlsCopy = np.copy(HLS_Img)
            l_portion = cv.getTrackbarPos("l", "l and s")
            s_portion = cv.getTrackbarPos("s", "l and s")
            # 调整饱和度和亮度
            hlsCopy[:, :, 1] = (1.0 + l_portion / float(MAX_VALUE)) * hlsCopy[:, :, 1]
            hlsCopy[:, :, 1][hlsCopy[:, :, 1] > 1] = 1
            hlsCopy[:, :, 2] = (1.0 + s_portion / float(MAX_VALUE)) * hlsCopy[:, :, 2]
            hlsCopy[:, :, 2][hlsCopy[:, :, 2] > 1] = 1

            lsImg = cv.cvtColor(hlsCopy, cv.COLOR_HLS2BGR)
            cv.imshow("LandS_image", lsImg)
            ch = cv.waitKey(10)
            if ch == 27:
                break

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print(f"Usage: python python-scripy.py imageFile.")

File: DigitalImageProcessing/cpp/convert_color.cpp
/**
 * @File    : convert_color.cpp
 * @Brief   : 色彩空间的转换; 调整彩色图像的饱和度和亮度
 * @Author  : Wei Li
 * @Date    : 2021-10-21
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

// CONST
cv::Mat image;
cv::Mat image_norm;
cv::Mat image_hls;
cv::Mat image_ls;

int width, height;
cv::String win_name = "Saturation_Liminance_Image";
int L = 0;
int S = 0;
int MAX_VALUE = 100;
void callBack_LS(int, void*);

// -------------------------------
int main(int argc, char** argv)
{
    if (argc > 1)
    {
        image = cv::imread(argv[1]);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }
        cv::imshow("OriginImage", image);

        width = image.cols;
        height = image.rows;
        image.convertTo(image_norm, CV_32FC3, 1.0 / 255, 0);
        cv::namedWindow(win_name, cv::WINDOW_AUTOSIZE);

        cv::createTrackbar("Saturation", win_name, &S, MAX_VALUE, callBack_LS);
        cv::createTrackbar("Liminance", win_name, &L, MAX_VALUE, callBack_LS);
        callBack_LS(0, 0);

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV python script imageFile." << std::endl;
        return -1;
    }
}


void callBack_LS(int, void*)
{
    cv::cvtColor(image_norm, image_hls, cv::COLOR_BGR2HLS);
    for (size_t r = 0; r < height; ++r)
    {
        for (size_t c = 0; c < width; ++c)
        {
            cv::Vec3f hls = image_hls.at<cv::Vec3f>(r, c);
            hls = cv::Vec3f(hls[0], 
                            (1 + L / double(MAX_VALUE))*hls[1] > 1 ? 1 : (1 + L / double(MAX_VALUE)) * hls[1],
                            (1 + S / double(MAX_VALUE))*hls[2] > 1 ? 1 : (1 + S / double(MAX_VALUE)) * hls[2]);
                            
        image_hls.at<cv::Vec3f>(r, c) = hls;
        }
    }

    cv::cvtColor(image_hls, image_ls, cv::COLOR_HLS2BGR);
    cv::imshow(win_name, image_ls);
}
